lxd/device/cdi/configure.go

```diff
@@ -0,0 +1,335 @@
+//     - The 'specific device' configuration: this is a list of 'container edits' that can be added to the container runtime.
+//     According to the CDI ID (vendor, class, name), we only select the 'container edits' that matches the CDI ID.
+//     The 'container edits' are a list of device nodes, hooks and mounts that must be added to the container runtime.
+//     - The 'general device' configuration: this is a single 'container edits' entry runtime that must be passed to the container runtime in ant case. Which unix char devices need to be passed
+//     (e.g, special GPU memory controller device, etc.)? Which user space libraries need to be mounted (e.g, CUDA libraries for NVIDIA, etc.)?
+//     Which hooks need to be executed (e.g, symlinks to create, folder entries to add to ldcache, etc.))?
+//     In our case, these edits will be interpreted either as disk or unix char mounts passed to the container.
+//     The hooks will be centralized in a single resource file that will be read and executed as a LXC `lxc.hook.mount` hook,
+//     through LXD's `callhook` command.
+//  2. We first process the 'specific device' configuration: we convert this information into a map of devices
+//     (keyed by their path given in the spec, it mapped to a map of device properties). We also collect the specific mounts (but we do not process them yet) and hooks.
+//  3. We then process the 'general device' configuration in the same fashion.
+//  4. Now we process all the mounts we collected from the spec in order to turn them into disk devices.
+//     This operations generate a side effect: it generates a list of indirect symlinks (see `specMountToNativeDev`)
+//  5. Merge all the hooks (direct + indirect) into a single list of hooks.
+func GenerateFromCDI(inst instance.Instance, cdiID ID, l logger.Logger) (*ConfigDevices, *Hooks, error) {
+	// 1. Generate the CDI specification
+	spec, err := generateSpec(cdiID, inst, l)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to generate CDI spec: %w", err)
+	}
+
+	// Initialize the hooks as empty
+	hooks := &Hooks{ContainerRootFS: inst.RootfsPath()}
+	mounts := make([]*specs.Mount, 0)
+	configDevices := &ConfigDevices{UnixCharDevs: make([]map[string]string, 0), BindMounts: make([]map[string]string, 0)}
+
+	// 2. Process the specific device configuration
+	for _, device := range spec.Devices {
+		if device.Name == cdiID.Name {
+			mounts, err = applyContainerEdits(device.ContainerEdits, configDevices, hooks, mounts, l)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			break
+		}
+	}
+
+	// 3. Process general device configuration
+	mounts, err = applyContainerEdits(spec.ContainerEdits, configDevices, hooks, mounts, l)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// CDI Mounts are converted to LXD `disk` devices
+	// If the instance is privileged, then we shift the disk mounts
```

Comment:
Why is shifting only needed if privileged?

Reply:
This is an other mistake: I initially thought that shifting a bind mount was only required for privileged container. I realize that it is not the case (we do this for LXD disk device without the container to be privileged...) . Then I'll enable shifting to true by default.

Reply:
fixed